apps/survey/contracts/Surveying.sol

```diff
@@ -0,0 +1,240 @@
+
+        uint256 i;
+        Survey storage survey = surveys[_surveyId];
+
+        uint256 voterStake = token.balanceOfAt(msg.sender, survey.snapshotBlock);
+        MultiOptionVote storage previouslyVoted = survey.voters[msg.sender];
+
+        // revert previous votes, if any
+        if (previouslyVoted.options.length > 0) {
+            // voter removes their vote
+            for (i = 1; i < previouslyVoted.options.length; i++) {
+                survey.votes[previouslyVoted.options[i]] = survey.votes[previouslyVoted.options[i]].sub(previouslyVoted.stakes[i]);
+            }
+            // remove previous vote participation
+            survey.participation = survey.participation.sub(voterStake.sub(previouslyVoted.stakes[0]));
+        }
+
+        // reset previously voted options, if any
+        delete survey.voters[msg.sender];
+
+        uint256 totalVoted = 0;
+        // reserve first index for NO_VOTE
+        survey.voters[msg.sender].options.push(NO_VOTE);
+        survey.voters[msg.sender].stakes.push(0);
+        for (i = 0; i < _options.length; i++) {
+            _voteOption(_surveyId, _options[i], _stakes[i]);
+            // let's avoid repeating an option by
+            // making sure that ascending order is preserved in options array
+            require(survey.voters[msg.sender].options[i] < survey.voters[msg.sender].options[i+1]);
+            totalVoted = totalVoted.add(_stakes[i]);
+        }
+
+        // compute and register non used tokens
+        // implictly we are doing require(totalVoted <= voterStake) too:
+        survey.voters[msg.sender].stakes[0] = voterStake.sub(totalVoted);
+
+        // add voter tokens to participation
+        survey.participation = survey.participation.add(totalVoted);
+    }
+
+    /**
+    * @notice Vote option #`_option` in survey #`_surveyId`.
+    * @dev It will use the whole balance. If voting option is 0, that's like cancelling the previously casted vote
+    * @param _surveyId Id for survey
+    * @param _option Index of supported option
+    */
+    function voteOption(uint256 _surveyId, uint256 _option) public {
+        Survey storage survey = surveys[_surveyId];
+        // this could re-enter, though we can asume the governance token is not maliciuous
+        uint256 voterStake = token.balanceOfAt(msg.sender, survey.snapshotBlock);
+        uint256[] memory options = new uint256[](1);
+        uint256[] memory stakes = new uint256[](1);
+        options[0]= _option;
+        stakes[0] = voterStake;
+
+        voteOptions(_surveyId, options, stakes);
+    }
+
+    function _voteOption(uint256 _surveyId, uint256 _option, uint256 _stake) public {
+        Survey storage survey = surveys[_surveyId];
+
+        require(_option != NO_VOTE && _option <= survey.options);
+        require(_stake > 0);
+
+
+        // register voter amount
+        survey.voters[msg.sender].options.push(_option);
+        survey.voters[msg.sender].stakes.push(_stake);
+
```

Comment:
The problem here was that the function was wrongly set as `public` instead of `internal`. With this function being internal I don't think a situation where those array lengths are different can be reached, right?